Add missing derives, `favicon` field and new `Packet` struct.

reitermarkus opened the pull request:
- Derive `Debug` for every struct.
- Add missing `favicon` field for `ServerStatus`.
- Add `Packet` struct to decode full packets.

reitermarkus committed:
Add missing derives.

reitermarkus committed:
Add `Packet` struct.

reitermarkus committed:
Add missing `favicon` field.

vaIgarashi commented on protocol/src/packet.rs:
```diff
@@ -0,0 +1,87 @@
+use crate::error::{DecodeError, EncodeError};
+
+#[derive(Debug)]
+pub struct Packet {
```
When you will intergrate protocol in a network library like tokio this structure would not be necessary. There are more steps in packet codec. Like encryption and compression.

reitermarkus replied:
> When you will intergrate protocol in a network library like tokio this structure would not be necessary.

Do you have an example for how this works?

I am currently converting a `tokio::net::TcpStream` into `std::net::TcpStream` so I can use `Packet::decode` with it, so this would definitely be an improvement over what I have now.

timvisee commented:
> Do you have an example for how this works?

I've bodged together packet reading/writing with compression support here:

https://github.com/timvisee/lazymc/blob/0c69752fecdc4ad95d2c5c96c3fd37f72ea11d70/src/proto/packet.rs#L18-L235

My implementation handles compression, but no encryption though.

In the login stage, the server may send a SetCompression(n) packet after which compression should be enabled if the packet payload is `>= n` bytes.

Read more about compression here: https://wiki.vg/Protocol#Packet_format

reitermarkus commented:
Thanks @timvisee, I will have a look, your “Lobby” feature looks very interesting.

For my use case ([a Kubernetes game server proxy for auto-scaling](https://github.com/reitermarkus/zero-scale-proxy), pretty similar to your `lazymc` project actually) I didn't need compression/encryption support since the only thing I need to know is whether a login packet has been received, after that everything else is forwarded as is to the upstream server.

reitermarkus commented:
Okay, looking at @timvisee's implementation, this `Packet` struct with its `Decoder` implementation could replace the `read_packet_id_data` function. Maybe calling it `UncompressedPacket` makes more sense then?

reitermarkus committed:
Support compression for packets.

reitermarkus commented:
Okay, I renamed the `Packet` to `RawPacket` and also added structs for `CompressedRawPacket` and `UncompressedRawPacket` to support both compressed and uncompressed packets.

reitermarkus committed:
Decode `CompressedRawPacket` to `UncompressedRawPacket`.

reitermarkus committed:
Add `Incomplete` error variant for incomplete packets.

reitermarkus committed:
Add `Packet` with custom `encode`/`decode` methods.

reitermarkus commented:
@vaIgarashi, can you have another look?

vaIgarashi commented on protocol/src/packet.rs:
```diff
@@ -0,0 +1,190 @@
+        if let Some(threshold) = compression_threshold {
+            CompressedRawPacket { packet, threshold }.encode(&mut buf)?;
+        } else {
+            packet.encode(&mut buf)?;
```
RawPacket missing encoder and decoder implementation?

vaIgarashi commented on protocol/src/packet.rs:
```diff
@@ -0,0 +1,190 @@
+            Err(DecodeError::IoError { io_error })
+                if io_error.kind() == io::ErrorKind::UnexpectedEof =>
+            {
+                return Err(DecodeError::Incomplete { bytes_needed: 1 })
```
This is not correct because read var int may vary from 1 to 4 bytes

reitermarkus replied:
This is used only as an indication, so it should be read as “at least 1”. If more are needed, it will happen in the next iteration.

reitermarkus replied:
Maybe it should be an `Option` so it works similar to https://docs.rs/nom/latest/nom/enum.Needed.html.

vaIgarashi commented on protocol/src/packet.rs:
```diff
@@ -0,0 +1,190 @@
+        let mut buf = Vec::new();
+        let packet = RawPacket {
+            id: self.id,
+            data: self.data.clone(),
```
Can we avoid data cloning?

reitermarkus replied:
Changed the method to take an owned `self`.

vaIgarashi commented on protocol/src/packet.rs:
```diff
@@ -0,0 +1,190 @@
+            id: self.id,
+            data: self.data.clone(),
+        };
+        if let Some(threshold) = compression_threshold {
```
nth: match are more preferable when you use else branch

vaIgarashi commented on protocol/src/packet.rs:
```diff
@@ -0,0 +1,190 @@
+            encoder.write_all(&packet_buf)?;
+            encoder.finish()?;
+        } else {
+            writer.write_var_i32(0)?;
```
Why zero?

reitermarkus replied:
https://wiki.vg/Protocol#With_compression

> If the size of the buffer containing the packet data and ID (as a VarInt) is smaller than the threshold specified in the packet Set Compression. It will be sent as uncompressed. This is done by setting the data length as 0. (Comparable to sending a non-compressed format with an extra 0 between the length, and packet data).

reitermarkus committed:
Avoid cloning.
